src/math/numbers.c

```diff
@@ -130,28 +130,16 @@ int32_t find_max_abs_int32(int32_t vec[], int vec_length)
  */
 int norm_int32(int32_t val)
 {
-	int s;
-	int32_t n;
-
-	if (!val)
-		return 31;
-
-	if (val > 0) {
-		n = val << 1;
-		s = 0;
-		while (n > 0) {
-			n = n << 1;
-			s++;
-		}
-	} else {
-		n = val << 1;
-		s = 0;
-		while (n < 0) {
-			n = n << 1;
-			s++;
-		}
-	}
-	return s;
+	int c = 0;
+
+	/* count number of bits c that val can be right-shifted arithmetically
+	 * until there is -1 (if val is negative) or 0 (if val is positive)
+	 * norm of val will be 31-c
+	 */
+	for (; val != -1 && val != 0; c++)
```

Comment:
What happens if val is smaller than 0 i.e. if (val<0) ?

Reply:
in case val is less than 0 and is right-shifted arithmetically, it will be sign-extended and eventually comes down to -1 (0xffffffff)
The number of bits c that negative val can be right-shifted until it becomes -1 constitutes a number of "useful bits" of val and norm will be 31-c.